doc/module-system/module-system.chapter.md

```diff
@@ -0,0 +1,69 @@
+
+### Parameters {#module-system-lib-evalModules-parameters}
+
+#### `modules` {#module-system-lib-evalModules-param-modules}
+
+A list of modules. These are merged together using various methods to form the final configuration.
+
+#### `specialArgs` {#module-system-lib-evalModules-param-specialArgs}
+
+An attribute set of module arguments that can be used in `imports`.
+
+This is in contrast to `config._module.args`, which is only available within the module fixpoint, which does not exist before all imports are resolved.
+
+#### `specialArgs.class` {#module-system-lib-evalModules-param-specialArgs-class}
+
+If the `class` attribute is set in `specialArgs`, the module system will rejected modules with a different `class`.
+
+This improves the error message that users will encounter when they import an incompatible module that was designed for a different class of configurations.
+
+The `class` value should be in camelcase, and, if applicable, it should match the prefix of the attributes used in (experimental) flakes. Some examples are:
+
+ - `nixos`: NixOS modules
+ - `nixosTest`: modules that constitute a [NixOS VM test](https://nixos.org/manual/nixos/stable/index.html#sec-nixos-tests)
+
+#### `prefix` {#module-system-lib-evalModules-param-prefix}
+
+A list of strings representing the location at or below which all options are evaluated. This is used by `types.submodule` to improve error reporting and find the implicit `name` module argument.
+
+### Return value {#module-system-lib-evalModules-return-value}
+
+#### `options` {#module-system-lib-evalModules-return-value-options}
+
+The nested set of all option declarations.
+
+#### `config` {#module-system-lib-evalModules-return-value-config}
+
+The nested set of all option values.
+
+#### `type` {#module-system-lib-evalModules-return-value-type}
+
+A module system type representing the module set as a submodule, to be extended by configuration from the containing module set.
+
+This is also available as the module argument `moduleType`.
+
+#### `extendModules` {#module-system-lib-evalModules-return-value-extendModules}
+
+A function similar to `evalModules` but building on top of the module set. Its arguments, `modules` and `specialArgs` are added to the existing values.
+
+Using `extendModules` a few times has no performance impact as long as you only reference the final `options` and `config`.
+If you do reference multiple `config` (or `options`) from before and after `extendModules`, performance is the same as with multiple `evalModules` invocations, because the new modules' ability to override existing configuration fundamentally requires a new fixpoint to be constructed.
+
+This is also available as a module argument.
+
+#### `_module` {#module-system-lib-evalModules-return-value-_module}
+
+A portion of the configuration tree which is elided from `config`. It contains some values that are mostly internal to the module system implementation.
```

Comment:
Should it be documented then? One could argue this shouldn't be exposed to begin with.

Reply:
This is tech debt until the markdown migration is complete and we can show the `_module` docs again.
Internal-ness has been debated.

Reply:
I don't think `_module` should be internal. As I can see, the main reason it should be prefixed with a `_` is to avoid clashing with users option names. In https://github.com/NixOS/nixpkgs/pull/97023 I even tried enabled rendering of these options in the manual. That entire PR was reverted, but that as a separate change would still be good.